### PR TITLE
feat/timezone_config

### PR DESCRIPTION
**Pull request by JarbasAl**
- all datetime extractors now default to `now_local` for anchorDate
- all `localized_function`s now inject a tzinfo in all tz-naive datetime objects passed as arguments
- default tz-info can be set with `lingua_franca.time.set_default_tz`
- tz-info injection can be enabled/disable in `lingua_franca.config`(default: `enabled`)

in what relates to mycroft-core:
- core should set the default timezone when setting language at intent step
- core should expect LF to always return tz-aware objects (in user configured timezone from .conf)

### CODE REVIEW

**Review by krisgesling — COMMENTED** [review]
I wanted to clarify the points in the PR description as that will likely change my inline comments below.

> - [x] all datetime extractors now default to now_local for anchorDate

This to me is the bug and what we're really trying to fix.

> all localized_functions now inject a tzinfo in all tz-naive datetime objects passed as arguments

This seems to be a feature made possible by the default tz config however I'm unclear as to how you are suggesting it work. Is it 1, 2, both, or something else?
1. LF users can pass in TZ naive datetime objects and they will have default tz info added.
2. LF users can choose to have all datetime objects converted to the default tz.

A possible issue arises in `nice_duration`. If I pass in 1 tz aware and 1 tz naive datetime, do I expect them both to be treated with the tzinfo of the tz aware datetime, or with my default tz info? I can see people making strong arguments in both directions, which to me adds weight to requiring that all datetime objects passed as parameters must be tz aware.

Enforcing the use of tz aware datetime objects reduces the chance of perceived bugs by warning downstream developers if they are using tz naive datetime objects. It also reduces the amount of datetime modifications we are doing.

> tz-info injection can be enabled/disable in lingua_franca.config(default: enabled)

For this I assume number 2 above, else why would it be disabled?

> default tz-info can be set with lingua_franca.time.set_default_tz

Wondering if this should get moved to `lingua_franca.config.set_default_tz`


> in what relates to mycroft-core:
> core should set the default timezone when setting language at intent step
> core should expect LF to always return tz-aware objects (in user configured timezone from .conf)

If the intention was only to fix the bug in `extract_datetime` then we could add a tz parameter to that method and have mycroft-core wrap it providing the backwards compatibility for all Skills effectively: 
`extract_datetime(text, anchorDate=None, default_tz=now_local(), default_time=None)`

I don't think we'd want to have all datetimes returned in local tz by default. However it seems we're only talking about `extract_datetime` that actually returns a datetime anyway right? In which case I would expect it be returned in the timezone of anchorDate or the default_tz if no anchorDate is provided.

File: lingua_franca/internal.py
Comment (krisgesling) [reply lingua_franca/internal.py:465]: This is probably covered better in the overall review comment.

File: lingua_franca/config.py
Comment (krisgesling) [reply lingua_franca/config.py:2]: This is probably covered better in the overall review comment.

File: lingua_franca/lang/parse_da.py
Comment (JarbasAl) [reply lingua_franca/lang/parse_da.py]: this was a copy paste mistake. 

tz needs to be attached here, but not converted. the original piece of code was assuming UTC before conversion but we dont want that here

File: lingua_franca/time.py
Comment (JarbasAl) [reply lingua_franca/time.py:25]: the possible formats are a tzinfo object, or a standard string for timezone, like the mycroft.conf has. those are pretty universal strings to identify timezones and i think most users will prefer using those instead of tzinfo objects. i don't think we want fuzzy matching at all in here

i missed this TODO, will add this mapping to the PR and a custom exception with logs about proper usage

File: lingua_franca/time.py
Comment (JarbasAl) [reply lingua_franca/time.py:23]: I'm not sure about that and i would like to tackle the whole config module in a different PR, me and @ChanceNCounter have been having some discussions on how config could look like

We may end up moving those, but I'd like to not do that for now since the config module is essentially a placeholder

File: lingua_franca/internal.py
Comment (JarbasAl) [reply lingua_franca/internal.py]: the idea is to convert only tz naive datetime objects, not convert every date.

good catch on this mistake, all unittests use args instead of kwargs so i missed this

File: lingua_franca/config.py
Comment (JarbasAl) [reply lingua_franca/config.py:2]: not setting the default_tz falls back to system tz or UTC as last resort. but internally it still ensures all datetime objects are tz aware

setting `inject_timezones = False` skips the tz attachment completely, so it is kinda like the old behavior (except for the `now_local` not being used bug). This makes the returned date timezone undefined/depend on input data like it currently is.

File: lingua_franca/time.py
Comment (ChanceNCounter) [reply lingua_franca/time.py:23]: #128 definitely wants to be more than variables in a module, that's for sure.

File: lingua_franca/time.py
Comment (JarbasAl) [inline lingua_franca/time.py:97]: this method had not been migrated from mycroft-core, added in order to deprecate the mycroft side function as described in https://github.com/MycroftAI/lingua-franca/issues/182#issuecomment-809776181

File: lingua_franca/time.py
Comment (chrisveilleux) [reply lingua_franca/time.py:23]: I agree with Gez on this one.  Why are we not just setting a config value?  I understand wanting to tackle it in a different PR.  Just curious the rationale for doing it this way.

**Review by chrisveilleux — CHANGES_REQUESTED** [review]

File: lingua_franca/lang/parse_nl.py
Comment (JarbasAl) [reply lingua_franca/lang/parse_nl.py:570]: because thats how it was in mycroft-core, #97 did a major cleanup of this, but theres still some places to go. essentially the original english parser used CamelCase and the rest just copy pasted it

File: lingua_franca/lang/parse_cs.py
Comment (JarbasAl) [reply lingua_franca/lang/parse_cs.py:769]: not in this PR, that just makes the commit history murky. theres a big rewrite #96 that hopefully will scale to all languages and help reducing the duplication, kinda like #42 

many of the languages are simply copy pasted from a similar one and adapted, that causes lots of code duplication. the initial func was huge to start with and just kept growing. we have been splitting this into smaller helper funcs over time, but different languages move at different rates

File: lingua_franca/time.py
Comment (JarbasAl) [reply lingua_franca/time.py:23]: a proper config does not exist yet... PRs should do one thing only, let's not add stuff to things we know will be deprecated very soon unless absolutely necessary, dont want to have to support backwards compatibility, so let's keep this to a minimum before anyone starts depending on it

ideally we will have a config object later, right now we should do things with wrapper methods. supporting variables at the top level of a module is not good moving forward!

File: lingua_franca/config.py
Comment (JarbasAl) [reply lingua_franca/config.py:2]: there is no config yet as said in previous comment, so timezone should not be defined here yet

by using wrapper methods is trivial to add a proper config later and point the method to it + update docs to deprecate method. adding a bunch of stuff to placeholder module that is essentially a name will likely cause trouble down the road.

for the purpose of this PR lets pretend config module does not exist yet, `inject_timezones = True` can be completely removed if you want, as mentioned it was mostly meant as a backwards compat flag and shouldnt really be set by end users 99.9% of the time

File: lingua_franca/lang/parse_cs.py
Comment (chrisveilleux) [reply lingua_franca/lang/parse_cs.py:769]: fair enough.

File: lingua_franca/config.py
Comment (JarbasAl) [reply lingua_franca/config.py:2]: sry, missed the main question, UTC vs system comes from `tzlocal() `documentation, it is there only for backwards compat and because thats the current mycroft-core default (if mycroft.conf is not set it uses `tzlocal()` )

File: lingua_franca/lang/parse_nl.py
Comment (chrisveilleux) [reply lingua_franca/lang/parse_nl.py:570]: Ok, guess we can table this for another time.

File: lingua_franca/time.py
Comment (chrisveilleux) [reply lingua_franca/time.py:23]: Resolving this conversation as there seems to be a consensus that this will be addressed in a later revision.

File: lingua_franca/config.py
Comment (chrisveilleux) [reply lingua_franca/config.py:2]: OK.  Resolving this conversation as it will be addressed in a later revision.

File: lingua_franca/internal.py
Comment (JarbasAl) [reply lingua_franca/internal.py:468]: as a general principle i agree, but a throw away variable inside of a loop i think its legit to be a single character, just like `i` is the most commonly used in `for` loops. usually when iterating `dict.items()` i see `k, v` used, stands for `key, value`. `idx` also is commonly used when using `enumerate`

if we know the type being returned `v` should always be descriptive, but it can be literally anything in this case

Comment (ChanceNCounter) [reply lingua_franca/internal.py:468]: I agree with specific respect to `i`/`j` for loops, and `k`/`v` keyvals, only because the latter is canonical Python, and the former is canonical everywhere.

In general I too agree that single-character variables suck, but I almost *expect* `for k, v in`. It's a meaningful unit of code =P

**Comment by ChanceNCounter** [comment]
@chrisveilleux I do see value in implementing this directly in `lingua_franca.config`, but I'm hesitant to extend the current config.py. #185 is currently roadmapped for the next "major" version (we're using minor numbers cuz "alpha") and it implements a proper config API.

The current `config` only contains one value, and it's just a top-level variable. The file wouldn't exist at all, except Neon and Chatterbox needed `config.load_langs_on_demand` or they couldn't upgrade. There was no reason to deny them that in the 0.3.0 upgrade, but the consequence is now there's that ugly *thing*.

As I see it, there are two options here:

* Use a setter function, as seen above, which will become redundant as soon as #185 is done
* Add another top-level variable `config.default_time_zone`, which will cease to exist as soon as #185 is done

One way or another it's something that implementations will have to change again in the future, but that's why it's `v0.x`. The difference is just how disruptive the change will be. If we go with the setter function, we can deprecate that function, and take as long as we want to do so, leaving implementations plenty of time to catch up. If we go with config.py, implementations will have to change it just to make the LF0.4 --> LF0.5 upgrade.

File: lingua_franca/internal.py
Comment (chrisveilleux) [reply lingua_franca/internal.py:468]: Is it really that hard to type the extra letters?  `for key, value`  If I get my way, every Mycroft repository will be running PyLint, which won't take too kindly to the one letter variables.  Another thing, usually more meaningful names can be used.  What does the key represent?  The value?  The code will be easier to read if there is some variable naming that helps with intent.  Sorry, fellas.  Not backing down on this one.

Comment (JarbasAl) [reply lingua_franca/internal.py:468]: would you like to open an issue against the python source code? theres over 100 pages of search results with this pattern

https://github.com/python/cpython/search?q=for

File: lingua_franca/internal.py
Comment (ChanceNCounter) [inline lingua_franca/internal.py:475]: 🤡

File: lingua_franca/internal.py
Comment (JarbasAl) [reply lingua_franca/internal.py]: at some point you should attempt to collaborate instead of "getting your way"

File: lingua_franca/internal.py
Comment (ChanceNCounter) [reply lingua_franca/internal.py:475]: Apparently there's no clown face, so I guess it's emoji math 😆 + 😒 🧀 + 🤖 = 🏅

File: lingua_franca/lang/parse_ca.py
Comment (JarbasAl) [reply lingua_franca/lang/parse_ca.py:340]: "Bikeshedding (also referred to as the law of triviality) describes a phenomenon where people spend a relatively large amount of time, energy, and other resources dealing with relatively minor issues.

The concept of bikeshedding has important implications, both in terms of personal productivity as well as in terms of group efficiency, since it can help identify situations where people waste valuable resources on trivial matters, while also neglecting important issues which require their attention."

File: lingua_franca/internal.py
Comment (krisgesling) [reply lingua_franca/internal.py:475]: I understand that there are different opinions on the use of single character variables particularly in contained loops however I don't think it is helpful to turn the PR into a joke. This seems to be the only reason for this PR not being merged now is that right?

Comment (JarbasAl) [reply lingua_franca/internal.py:475]: the joke is how this PR was handled

if this was blocking the PR getting merged noone made any comment about it

if code style is blocking this PR , then i remind that this change was explicitly requested in opposition to what the repo maintainers (ex-maintainer in my case) wanted, so don't come and tell me the PR is blocked because of this change.... this PR is not merged because the maintainers are not allowed to maintain the repo

**Pull request closed without merging**
